File: frontend/ErasmusExpenseTracker-front/src/models/Transaction.js
export default class Transaction {
  constructor({
    transactionId,
    type,
    amount,
    currency,
    category,
    date,
    description,
    user,
    trip
  }) {
    this.transactionId = transactionId;
    this.type = type; // 'INCOME' o 'EXPENSE'
    this.amount = parseFloat(amount);
    this.currency = currency;
    this.category = category;
    this.date = new Date(date);
    this.description = description || '';
    this.user = user || null; // Puede ser un ID o un objeto User
    this.trip = trip || null; // Puede ser un ID o un objeto Trip
  }

  isIncome() {
    return this.type === 'INCOME';
  }

  isExpense() {
    return this.type === 'EXPENSE';
  }

  formattedAmount() {
    return `${this.amount.toFixed(2)} ${this.currency}`;
  }
}

File: backend/src/main/java/com/eet/backend/model/Transaction.java
package com.eet.backend.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.GenericGenerator;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transaction {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "transaction_id", updatable = false, nullable = false)
    private UUID transactionId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;

    @Column(nullable = false, precision = 19, scale = 4)
    private BigDecimal amount;

    @Column(nullable = false)
    private String currency;

    @Column(nullable = false)
    private LocalDate date;

    private String description;

    // Relaciones
    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    @JsonIgnore
    private User user;

    @ManyToOne
    @JoinColumn(name = "trip_id")
    @JsonIgnore
    private Trip trip;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

}

